x/storage/keeper/cross_app_permission.go

```diff
@@ -146,16 +147,88 @@ func (app *PermissionApp) handleCreatePolicySynPackage(ctx sdk.Context, createPo
-			Payload: types.CreatePolicyAckPackage{
-				Status:    types.StatusFail,
-				Creator:   createPolicyPackage.Operator,
-				ExtraData: createPolicyPackage.ExtraData,
-			}.MustSerialize(),
-			Err: err,
+		var msgPutPolicy types.MsgPutPolicy
+		err = msgPutPolicy.Unmarshal(createPolicyPackage.Data)
+		if err != nil {
+			return sdk.ExecuteResult{
+				Payload: types.CreatePolicyAckPackage{
+					Status:    types.StatusFail,
+					Creator:   createPolicyPackage.Operator,
+					ExtraData: createPolicyPackage.ExtraData,
+				}.MustSerialize(),
+				Err: err,
+			}
+		}
+		if msgPutPolicy.GetOperator() != createPolicyPackage.Operator.String() {
+			return sdk.ExecuteResult{
+				Payload: types.CreatePolicyAckPackage{
+					Status:    types.StatusFail,
+					Creator:   createPolicyPackage.Operator,
+					ExtraData: createPolicyPackage.ExtraData,
+				}.MustSerialize(),
+				Err: types.ErrAccessDenied.Wrapf(
+					"Only resource owner can put policy, operator (%s)",
+					msgPutPolicy.GetOperator()),
+			}
+		}
+
+		err = grn.ParseFromString(msgPutPolicy.Resource, true)
+		if err != nil {
+			return sdk.ExecuteResult{
+				Payload: types.CreatePolicyAckPackage{
+					Status:    types.StatusFail,
+					Creator:   createPolicyPackage.Operator,
+					ExtraData: createPolicyPackage.ExtraData,
+				}.MustSerialize(),
+				Err: err,
+			}
 		}
+
+		if msgPutPolicy.ExpirationTime != nil && msgPutPolicy.ExpirationTime.Before(ctx.BlockTime()) {
+			return sdk.ExecuteResult{
+				Payload: types.CreatePolicyAckPackage{
+					Status:    types.StatusFail,
+					Creator:   createPolicyPackage.Operator,
+					ExtraData: createPolicyPackage.ExtraData,
+				}.MustSerialize(),
+				Err: permtypes.ErrPermissionExpired.Wrapf("The specified policy expiration time is less than the current block time, block time: %s", ctx.BlockTime().String()),
+			}
+		}
+
+		for _, s := range msgPutPolicy.Statements {
+			if s.ExpirationTime != nil && s.ExpirationTime.Before(ctx.BlockTime()) {
+				return sdk.ExecuteResult{
+					Payload: types.CreatePolicyAckPackage{
+						Status:    types.StatusFail,
+						Creator:   createPolicyPackage.Operator,
+						ExtraData: createPolicyPackage.ExtraData,
+					}.MustSerialize(),
+					Err: permtypes.ErrPermissionExpired.Wrapf("The specified statement expiration time is less than the current block time, block time: %s", ctx.BlockTime().String()),
+				}
+			}
+		}
+
+		policy = permtypes.Policy{
+			ResourceType:   grn.ResourceType(),
+			Principal:      msgPutPolicy.Principal,
+			Statements:     msgPutPolicy.Statements,
+			ExpirationTime: msgPutPolicy.ExpirationTime,
+		}
+		_, resID, err := app.getResourceOwnerAndIdFromGRN(ctx, grn)
+		if err != nil {
+			return sdk.ExecuteResult{
+				Payload: types.CreatePolicyAckPackage{
+					Status:    types.StatusFail,
+					Creator:   createPolicyPackage.Operator,
+					ExtraData: createPolicyPackage.ExtraData,
+				}.MustSerialize(),
+				Err: err,
+			}
+		}
+		policy.ResourceId = resID
```

Comment:
To me, this feature is to support resource name, so this assign is the key point. 
But before this, a lot of code is added to do the verification, what is the motivation? Is the previous implementation missed to do verification?

Reply:
accept,fixed